src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs

```diff
@@ -89,8 +90,19 @@ static MsQuicApi()
 
         if (OperatingSystem.IsWindows())
         {
-            // Windows ships msquic in the assembly directory.
-            loaded = NativeLibrary.TryLoad(Interop.Libraries.MsQuic, typeof(MsQuicApi).Assembly, DllImportSearchPath.AssemblyDirectory, out msQuicHandle);
+            // we ship Schannel version of MsQuic as part of the .NET runtime.
+            // This version works on newer OS versions (see s_minWindowsVersion).
+            //
+            // For cases where the Schannel version cannot be used, we want to
+            // support developers explicitly providing OpenSSL version of MsQuic.
+            // in the application directory, so we first check there and default
+            // to the Schannel version if not found.
+            loaded = NativeLibrary.TryLoad(Path.Combine(AppContext.BaseDirectory, Interop.Libraries.MsQuic), out msQuicHandle);
```

Comment:
I tried to use variations of

```cs
loaded = NativeLibrary.TryLoad(Interop.Libraries.MsQuic, Assembly.GetEntryAssembly(), DllImportSearchPath.ApplicationDirectory, out msQuicHandle);
```cs

But the loading procedure still seemed to prefer the msquic.dll next to the runtime assemblies. Not sure if that is expected behavior. cc @jkotas who might know.

Reply:
cc: @janvorli @elinor-fung @vitek-karas

Reply:
For reference, the default native library loading algorithm is documented at https://learn.microsoft.com/en-us/dotnet/core/dependency-loading/default-probing#unmanaged-native-library-probing . It does not let you control over resolution of name conflicts.